### ISSUE TITLE
Line height is distributed unevenly when lineHeight <= fontSize

### ISSUE THREAD

**Issue by andymatuschak** [facebook/react-native#29507]
## Description
When `lineHeight` is less than the `fontSize`, `Text` shrinks the line's bounding box only by removing space from above the text, rather than distributing the space evenly above and below, as is done with extra line height. Actually, the issue manifests when `lineHeight` is less than `fontSize` plus some small amount (perhaps the font bounding box height?)—e.g. <= 45px for 40px Helvetica.

This leads to clipped glyphs and produces differences in text rendering between react-native-web and react-native (https://github.com/necolas/react-native-web/issues/1687). After some discussion with @necolas, my inclination is that RNW's behavior is the more reasonable one, so I thought I'd open an issue here.

This issue reproduces across a variety of fonts.

![image](https://user-images.githubusercontent.com/2771/88598530-50b72e80-d01e-11ea-8dac-6d712e667e38.png)

Unfortunately,I worry that fixing this issue will create subtle and surprising source compatibility issues for existing clients. 
## React Native version:
```
System:
    OS: macOS 10.15.5
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 337.02 MB / 32.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.21.0 - ~/.nvm/versions/node/v10.21.0/bin/node
    Yarn: 1.19.1 - /usr/local/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v10.21.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.9.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 13.6, DriverKit 19.0, macOS 10.15, tvOS 13.4, watchOS 6.2
    Android SDK: Not Found
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.5900203
    Xcode: 11.6/11E708 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.2 - /usr/bin/javac
    Python: 2.7.16 - /usr/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: Not Found
    react-native: 0.63.2 => 0.63.2 
  npmGlobalPackages:
    *react-native*: Not Found
```

## Steps To Reproduce
Run example app here: https://snack.expo.io/W51F2OAqD

## Expected Results
The extra line spacing should be removed evenly from both above and below the type.

## Snack, code example, screenshot, or link to a repository:
https://snack.expo.io/W51F2OAqD

**Comment by andymatuschak** [comment]
A workaround for readers who stumble on this issue: if you'd like to use smaller line heights without glyph clipping, you'll need to add top padding to the `Text` element, which you must then compensate for by shifting the element up (e.g. via `top` plus relative positioning, or a negative `marginTop`).

**Comment by ravirajn22** [comment]
Does adding this prop help `textAlignVertical='center'`?

**Comment by andymatuschak** [comment]
@ravirajn22: Nope, that doesn't change the output, neither on Android nor on iOS.

**Comment by andymatuschak** [comment]
Nope, the issue has not been fixed; it still requires attention. I don't seem to have permission to add it to the backlog.

**Comment by Dinkelborg** [comment]
I can confirm that this is still an issue

**Comment by SpookyUmi** [comment]
Yep, still an issue

**Comment by navaneeth7067** [comment]
Still an issue

**Comment by kbrandwijk** [comment]
Agreed, still an issue, bot.

**Comment by fmacedoo** [comment]
Hey, still an issue! :D

**Comment by aschultz** [comment]
On Windows, XAML has [LineStackingStrategy](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.LineStackingStrategy?view=winrt-22000) and [TextLineBounds](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.textlinebounds?view=winrt-22000) for fine-tuning the text positioning. On web, there's [a proposal](https://www.w3.org/TR/css-inline-3/#propdef-leading-trim) for "[leading-trim](https://medium.com/microsoft-design/leading-trim-the-future-of-digital-typesetting-d082d84b202)". 

Do iOS or Android have anything similar that could be used or exposed?

**Comment by v0vchansky** [comment]
Still an issue

**Comment by sungsong88** [comment]
Same issue here. Basically, any characters that were not part of the font package were causing this issue for me.

**Comment by elencho** [comment]
https://github.com/facebook/react-native/issues/7687

**Comment by alexi-stein** [comment]
Definitely still an issue and the workaround is not great if you ever have Text within Text, they no longer align when they both have some kind of top offset

**Comment by HaakonSvane** [comment]
Would be great to bring some attention to this issue. The workaround mentioned in #7687 is hacky and will not work properly with special capital characters (e.g. Å, Á, È) unless you manually tweak paddings to "just about right".

**Comment by Mr-Fullstack** [comment]
try verticalAlign:"bottom"

**Comment by jeonghoonkim** [comment]
Still an issue

**Comment by fabriziobertoglio1987** [comment]
it is caused by the baseline of the Text. The Text and TextInput are similar, you can find relevant explanations of how baseline is set on react native in this prs https://github.com/facebook/react-native/pull/38359 https://github.com/facebook/react-native/pull/37465. I'll follow up

**Comment by oliviercperrier** [comment]
Still an issue

**Comment by OlacodesLA** [comment]
Still an issue

**Comment by keithpickering** [comment]
To add to the pile: this is still an issue. The paddingTop hack works for `Text` components, but not for `TextInput` (the blinking cursor is going to be as tall as whatever imaginary line height iOS decided to make up)

React Native supports a web-like styling paradigm, so the rendering should be web-like too. Hope we can get this fixed at some point.

Are there any known workarounds for Text inputs at this time?

**Comment by fabOnReact** [comment]
the fix is here https://github.com/facebook/react-native/pull/44614

**Comment by lsarni** [comment]
It is also a problem when using special characters even if the lineHeight > fontSize as they get cut off. The hacky padding workaround works but is not ideal https://github.com/facebook/react-native/issues/29507#issuecomment-1529425022

Here is `lineHeight: 37` and `fontSize: 32`
Android
![image](https://github.com/user-attachments/assets/cd537903-ac9e-4d6b-9ff5-4edf1069ac2c)
Web
![image](https://github.com/user-attachments/assets/6aa33826-f9ea-4092-92a1-6d454182f04d)

https://snack.expo.dev/@lsarni/petrified-kiwi

**Comment by ManuLpz4** [comment]
This is crazy, just to follow up it

**Comment by mellyeliu** [comment]
fyi: we've added a fix for this here: https://github.com/facebook/react-native/pull/46362, although we are still testing before rollout

**Comment by cortinico** [comment]
Closing as the PR mentioned by @mellyeliu has been merged and will be shipped with React Native 0.76

**Comment by DimitarNestorov** [comment]
@mellyeliu @cortinico The issue seems to affect both Android and iOS and the PR only introduces a solution for Android. Can we reopen until iOS is also addressed?

**Comment by mellyeliu** [comment]
Reopening until iOS fix for tracking purposes

**Comment by ArekChr** [comment]
@mellyeliu I have created a [PR](https://github.com/facebook/react-native/pull/46884) with a fix for iOS. Please take a moment to review It. Your feedback would be greatly appreciated!

**Comment by DimitarNestorov** [comment]
Not stale. PR #46884 is open

**Comment by DimitarNestorov** [comment]
Can you remove https://github.com/facebook/react-native/labels/Resolution%3A%20Fixed label as the fix for iOS is not merged yet?

**Comment by harmony-netizen** [comment]
Very surprised that such a fundamental issue for such a widely used framework is still unresolved for so long.
Am new to developing with React Native and ran into this issue from the very beginning.

The solution suggested by https://github.com/facebook/react-native/issues/29507#issuecomment-665147452 isn't suitable for my use case as I deal with dynamic content throughout the app with many different font sizes and some are multiline.

Mainly becomes an issue when the text is aligned with icons or there are line separators between lines of text where you can notice the uneven spacing.

Only solution I found was to increase the line height to at least 1.5 times the font size, but it doesn't always work for all cases (some just still look unbalanced regardless).
Or to reduce the icon size to give the illusion that it's centered.

Really wish that more attention would be paid to this issue.

**Comment by Nader-CS** [comment]
this issue still exist

**Comment by leineveber** [comment]
the android fix was merged but text is still cut off and the old padding hack doesnt work in new arch so bump

**Comment by leineveber** [comment]
@fabOnReact I saw in different threads that you primarily work on Text components, do u have any ideas on how to solve text cut offs for android (except of adjusting lh)? for ios its still possible to do padding + negative margin while for android it's impossible in new arch

jic example of cut off in the screenshot (fz = lh, first text is Noto Sans JP, second text is MPlus 1c):
<img width="685" height="345" alt="Image" src="https://github.com/user-attachments/assets/a6640c0f-7878-40c6-a642-8db0d11b8e60" />
